### ISSUE TITLE
GitHub API Integration

### ISSUE THREAD

**Issue by jaswsinc** [websharks/wp-kb-articles#2]
We have quite a bit of work to do when it comes to a GitHub integration. However, I'm going to open this issue for the purpose of establishing one class that we can start on right away.

- [x] We need a class responsible for loading and connecting to the GitHub API.
 - [x] This class should load the GitHub SDK. Ideally, this will either use transport layers already established by the GitHub SDK (e.g. cURL, FOPEN) or be a smaller custom class that integrates with the `WP_Http` class in WordPress. Either of these approaches would be fine. Whatever makes the most sense; i.e. whatever will make this quick, easy, and the most reliable. I suspect that will be one of the [PHP SDKs provided by GitHub](https://developer.github.com/libraries/), which we could include as a submodule.
  - [x] Method(s) needed to navigate an entire Git Data tree, even if it requires multiple API calls to follow any pagination links required. We need this to return a full overview of the entire tree.
  - [x] Method(s) needed to read a raw file from the Git Data tree.
  - [x] Method(s) needed to read [YAML](http://www.yaml.org/start.html) Headers from a file in the Git Data tree.

**YAML Headers:** This is to help clarify how the YAML Headers will be used in KB articles. KB articles will be written in Markdown, and Headers can be written in YAML at the top of the file, encapsulated by `---`.

```
---
title: A Quick Example KB Article
categories: documentation, how-tos, shortcodes
tags: shortcodes, file downloads, amazon s3, cloudfront
slug: s2-stream-shortcode
status: draft|publish
pubdate: [strtotime compatible]
excerpt: [short description/summary]
comments: on|off|true|false|0|1
author: jaswsinc
---

Article content
```

Another example of this can be found in GitHub Pages, which uses "Front-matter with Jekyll". The example below follows the same syntax. http://jekyllrb.com/docs/frontmatter/

---

**NOTE:** It's important to make this a generic class that is highly focused on the details above, and only on those details. Let's design this so that it will be a piece of the larger picture; i.e. there's no need to deal with any CRON jobs here, or make an attempt to integrate this with WP posts. Those items will be covered by other issues and from a different class file altogether.

In short, this class should be responsible for reading and parsing data from the Git Data API in the ways that are required to complete the larger project, but be limited to that functionality; i.e. not try to incorporate anything other than what I listed above.

---

Any values that should be pulled from a config file should exist as configurable class properties; i.e. making it possible to instantiate the class with a certain set of credentials, or with a certain repo URL, etc. Any of these should be a part of the class constructor.
